### DISCUSSION TITLE
[mini.files] Auto-expand empty & nested dirs

### DISCUSSION THREAD

**Discussion by iparnelly** [nvim-mini/mini.nvim#1184] [Show and tell]
I wrote a quick autocommand to automatically traverse directories that are empty outside of a single nested directory.

https://github.com/user-attachments/assets/08ce054e-c92e-4904-ac5b-e771c59ded01

Directory Structure for demo above:
```sh
> tree

.
├── README.md
└── a
    └── b
        └── c
            ├── d
            │   └── e
            │       └── node.lua
            └── node.lua

6 directories, 3 files
```

### Implementation
I basically just check if there is only 1 line in the buffer, and confirm that the entry is a directory. If it is, then I call `go_in`. 

I'm not lua-fluent (or even vim-api fluent) so if there are any issues you see that this might run into, let me know! 

I ran into a nice recursive loop situation, hence the `vim.schedule`. Otherwise, things are pretty straightforward.

### Code:
```lua
vim.api.nvim_create_autocmd({ "User" }, {
  pattern = { "MiniFilesWindowOpen" },
  group = require("config.autocmds").create_augroup_name("minifiles_auto_expand"),
  callback = function(data)
    local bufnr = data.data.buf_id
    local line_count = vim.api.nvim_buf_line_count(bufnr)
    if line_count ~= 1 then
      return
    end

    local entry = require("mini.files").get_fs_entry(bufnr, 1)
    if entry and entry.fs_type == "directory" then
      vim.schedule(function()
        require("mini.files").go_in()
      end)
    end
  end,
})
```

Thanks

**Comment by echasnovski** [comment]
Thanks for sharing! This looks nice!

The way I'd approach this, though, is via custom mapping. This looks more robust than `MiniFilesWindowOpen` autocommand (mainly because it will also work when windows are still open). So something like this:

```lua
local expand_single_dir
expand_single_dir = vim.schedule_wrap(function()
  local is_one_dir = vim.api.nvim_buf_line_count(0) == 1 and (MiniFiles.get_fs_entry() or {}).fs_type == 'directory'
  if not is_one_dir then return end
  MiniFiles.go_in()
  expand_single_dir()
end)

local go_in_and_expand = function()
  MiniFiles.go_in()
  expand_single_dir()
end

vim.api.nvim_create_autocmd('User', {
  pattern = 'MiniFilesBufferCreate',
  callback = function(args) vim.keymap.set('n', 'l', go_in_and_expand, { buffer = args.data.buf_id }) end,
})

require('mini.files').setup({ mappings = { go_in = '' } })
```

Mapping's LHS can be change to be something other than `l`, which is mostly a matter of taste here.

**Reply by iparnelly** [reply]
Hey thanks for the feedback! 

I didn't follow this though, can you expand?
> mainly because it will also work when windows are still open




I'll take your word for it though. I ended up modifying your `go_in_and_expand` function to allow for passing in args to `go_in()` (e.g. `{close_on_true}`)

```lua
local go_in_and_expand = function(...)
  local fs_entry = MiniFiles.get_fs_entry()
  if not fs_entry then
    return
  end

  MiniFiles.go_in(...)
  if fs_entry.fs_type == "directory" then
    expand_single_dir()
  end
end

vim.api.nvim_create_autocmd('User', {
  pattern = 'MiniFilesBufferCreate',
  callback = function(args) 
    vim.keymap.set("n", "l", function()
      go_in_and_expand({ close_on_file = true })
    end, { buffer = args.data.buf_id })
    vim.keymap.set("n", "L", function()
      go_in_and_expand()
    end, { buffer = args.data.buf_id })
  end,
})
```

**Reply by echasnovski** [reply]
> I didn't follow this though, can you expand?

On what exactly? It creates a buffer-local mapping for every buffer inside 'mini.files' explorer. When that mapping is executed, it calls `go_in()` followed by a recursive "expand single directory" action. This is both more straightforward to reason about (for example, because it is executed on demand and not as a side effect of an autocommand) and should work in more cases (like when there is already an open window to the right which contains single directory; it will be skipped with this version and not skipped with `MiniFilesWindowOpen` autocommand).
